hi-ml/src/health_ml/run_ml.py

```diff
@@ -190,48 +316,38 @@ def run(self) -> None:
 
-    def run_inference(self, checkpoint_path: Path) -> None:
-        """
-        Run inference on the test set for all models.
+        # do training
+        with logging_section("Model training"):
+            self.run_training()
 
-        :param checkpoint_path: The path to the checkpoint that should be used for inference.
-        """
-        lightning_model = self.container.model
-        if type(lightning_model).test_step != LightningModule.test_step:
-            # Run Lightning's built-in test procedure if the `test_step` method has been overridden
-            logging.info("Running inference via the LightningModule.test_step method")
-            # Lightning does not cope with having two calls to .fit or .test in the same script. As a workaround for
-            # now, restrict number of GPUs to 1, meaning that it will not start DDP.
-            self.container.max_num_gpus = 1
-            # Without this, the trainer will think it should still operate in multi-node mode, and wrongly start
-            # searching for Horovod
-            if ENV_OMPI_COMM_WORLD_RANK in os.environ:
-                del os.environ[ENV_OMPI_COMM_WORLD_RANK]
-            # From the training setup, torch still thinks that it should run in a distributed manner,
-            # and would block on some GPU operations. Hence, clean up distributed training.
-            if torch.distributed.is_initialized():  # type: ignore
-                torch.distributed.destroy_process_group()  # type: ignore
+        # load model checkpoint for custom inference or additional validation step
+        if self.container.has_custom_test_step() or self.container.run_extra_val_epoch:
+            self.load_model_checkpoint_after_training()
 
-            trainer, _ = create_lightning_trainer(self.container, num_nodes=1)
+        # Run extra validation epoch if enabled
+        if self.container.run_extra_val_epoch:
+            with logging_section("Model Validation to save plots on validation set"):
+                self.run_validation()
 
-            self.container.load_model_checkpoint(checkpoint_path=checkpoint_path)
-            data_module = self.container.get_data_module()
+        # Kill all processes besides rank 0
+        self.after_ddp_cleanup(old_environ)
 
-            # Change to the outputs folder so that the model can write to current working directory, and still
-            # everything is put into the right place in AzureML (there, only the contents of the "outputs" folder
-            # retained)
-            with change_working_directory(self.container.outputs_folder):
-                _ = trainer.test(self.container.model, datamodule=data_module)
+        # Run inference on a single device
+        with logging_section("Model inference"):
+            self.run_inference()
 
-        else:
-            logging.warning("None of the suitable test methods is overridden. Skipping inference completely.")
+        with logging_section("Regression Test"):
```

Comment:
Maybe this `logging_section` should be inside of the function, so that the "starting regression test" header is only output when we actually find a regression test folder?

Reply:
That's a good point, will change it.